src/composer/verifier.rs

```diff
@@ -0,0 +1,227 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use alloc::vec::Vec;
+use core::marker::PhantomData;
+
+use dusk_bls12_381::BlsScalar;
+use dusk_bytes::{DeserializableSlice, Serializable};
+use merlin::Transcript;
+
+use crate::commitment_scheme::OpeningKey;
+use crate::error::Error;
+use crate::proof_system::{Proof, VerifierKey};
+use crate::transcript::TranscriptProtocol;
+
+use super::Builder;
+
+/// Verify proofs of a given circuit
+pub struct Verifier<C> {
+    label: Vec<u8>,
+    verifier_key: VerifierKey,
+    opening_key: OpeningKey,
+    public_input_indexes: Vec<usize>,
+    transcript: Transcript,
+    size: usize,
+    constraints: usize,
+    circuit: PhantomData<C>,
+}
+
+impl<C> Verifier<C> {
+    pub(crate) fn new(
+        label: Vec<u8>,
+        verifier_key: VerifierKey,
+        opening_key: OpeningKey,
+        public_input_indexes: Vec<usize>,
+        size: usize,
+        constraints: usize,
+    ) -> Self {
+        let transcript =
+            Transcript::base(label.as_slice(), &verifier_key, constraints);
+
+        Self {
+            label,
+            verifier_key,
+            opening_key,
+            public_input_indexes,
+            transcript,
+            size,
+            constraints,
+            circuit: PhantomData,
+        }
+    }
+
+    fn prepare_serialize(
+        &self,
+    ) -> (usize, [u8; VerifierKey::SIZE], [u8; OpeningKey::SIZE]) {
+        let verifier_key = self.verifier_key.to_bytes();
+        let opening_key = self.opening_key.to_bytes();
+
+        let label_len = self.label.len();
+        let verifier_key_len = verifier_key.len();
+        let opening_key_len = opening_key.len();
+        let public_input_indexes_len = self.public_input_indexes.len() * 8;
+
+        let size = 48
+            + label_len
+            + verifier_key_len
+            + opening_key_len
+            + public_input_indexes_len;
+
+        (size, verifier_key, opening_key)
+    }
+
+    /// Serialized size in bytes
+    pub fn serialized_size(&self) -> usize {
+        self.prepare_serialize().0
```

Comment:
Same as in prover: I would calculate the size here and not in `prepare_serialize` so that we don't do unnecessary work when calling for the size

Reply:
Ditto; lets keep the discussion centralized there